sdk/search/Azure.Search.Documents/src/SearchServiceClient.cs

```diff
@@ -261,30 +256,309 @@ public virtual SearchIndexClient GetSearchIndexClient(string indexName)
+                cancellationToken);
+
+        /// <summary>
+        /// Creates a new data source.
+        /// </summary>
+        /// <param name="dataSource">Required. The <see cref="DataSource"/> to create.</param>
+        /// <param name="options">Optional <see cref="SearchRequestOptions"/> to customize the operation's behavior.</param>
+        /// <param name="cancellationToken">Optional <see cref="CancellationToken"/> to propagate notifications that the operation should be canceled.</param>
+        /// <returns>
+        /// The <see cref="Response{T}"/> from the server containing the <see cref="DataSource"/> that was created.
+        /// This may differ slightly from what was passed in since the service may return back properties set to their default values.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="dataSource"/> is null.</exception>
+        /// <exception cref="RequestFailedException">Thrown when a failure is returned by the Search service.</exception>
+        [ForwardsClientCalls]
+        public virtual async Task<Response<DataSource>> CreateDataSourceAsync(
+            DataSource dataSource,
+            SearchRequestOptions options = null,
+            CancellationToken cancellationToken = default) =>
+            await DataSourcesClient.CreateAsync(
+                dataSource,
+                options?.ClientRequestId,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        /// <summary>
+        /// Creates a new data source or updates an existing data source.
+        /// </summary>
+        /// <param name="dataSource">Required. The <see cref="DataSource"/> to create or update.</param>
+        /// <param name="accessConditions">Optional match conditions used to determine whether the data source should be updated based on whether it changed.</param>
+        /// <param name="options">Optional <see cref="SearchRequestOptions"/> to customize the operation's behavior.</param>
+        /// <param name="cancellationToken">Optional <see cref="CancellationToken"/> to propagate notifications that the operation should be canceled.</param>
+        /// <returns>
+        /// The <see cref="Response{T}"/> from the server containing the <see cref="DataSource"/> that was created.
+        /// This may differ slightly from what was passed in since the service may return back properties set to their default values.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="dataSource"/> is null.</exception>
+        /// <exception cref="RequestFailedException">Thrown when a failure is returned by the Search service.</exception>
+        [ForwardsClientCalls]
+        public virtual Response<DataSource> CreateOrUpdateDataSource(
+            DataSource dataSource,
+            MatchConditions accessConditions = null,
+            SearchRequestOptions options = null,
```

Comment:
We might consider creating an Options class that unions these together for this operation.  We can address that comprehensively in the future.

Reply:
It had crossed my mind. I'll open a bug to consider.

Reply:
Opened #11052 to track.